proposals/sycl_modules.md

```diff
@@ -700,73 +730,197 @@ module.
+Specialization constants are associated with the SYCL 1.2.1 program class,
 and named using a C++ type like kernels.
-The program class gains a `set_spec_constant` method, that sets a runtime
+The program class gains a `set_specialization_constant` method, that sets a runtime
 value to the specialization constant.
-The program can then be build with `build_with_kernel_type` to create a 
+The program can then be build with `build_with_kernel_type` to create a
 specialized kernel.
 
+### Specialization constant types
+
 On the new, type-safe, module approach, specialization constants are 
-associated with the OpenCL backend. 
-Other backends (e.g. Vulkan) may offer the same functionality, so the
-specialization constant type is defined as an (experimental) general 
-SYCL type as below:
+typically associated with SPIR-V module images but are not limited to such images.
+The specialization constant type is defined as a general SYCL type as below:
 
 ```cpp
 namespace sycl {
-namespace experimental {
 
-template <typename T, typename ID = T>
-class spec_constant {
+template <typename T>
+class specialization_id {
 private:
   // Implementation defined constructor.
-  spec_constant(/* Implementation defined */);
+  specialization_id(const specialization_id&) = delete;
+  specialization_id(specialization_id&&) = delete;
+
+
+public:
+
+  using type = T;
+
+  // Argument `Args` are forwarded to an underlying T Ctor.
+  // This allow the user to setup a default value for the specialization_id instance.
+  // The initialization of T must be evaluated at compile time to be valid.
+  template<class... Args >
+  explicit constexpr specialization_id(Args&&...);
+};
+
+template <class T, specialization_id<T>& s>
+class specialization_constant {
 public:
-  spec_constant();
+  using type = T;
+
+  // Create an empty (invalid) specialization constant.
+  // Remains invalid until initialized by a call
+  // to handler::set_specialization_constant or
+  // to handler::get_specialization_constant
+  specialization_constant();
+
+  specialization_constant(const specialization_constant&) = default;
+  specialization_constant& operator=(const specialization_constant&) = default;
 
   T get() const; // explicit access.
   operator T() const;  // implicit conversion.
 };
 
-}  // namespace experimental
 }  // namespace sycl
 ```
 
-`spec_constant` is a *placeholder type* used by the *SYCL implementation* to
+`specialization_id` is used by the *SYCL implementation* to *reference* constant whose
+concrete value will only be known during the application execution.
+To create a new `specialization_id` in a module, the user creates a `specialization_id` object with an *automatic* or *static* storage duration in the namespace or class scope.
+The user then uses references to this object to bind the `specialization_constant` value to the *identifier*.
+
+`specialization_id` objects must be forward declarable and cannot be moved or copied.
```

Comment:
Now that I understand that you are interested in the address of the object and not the value of the object, I need to think about how it works with `extern` and multiple translation units.
The object could be moved or copied but at the end it is just another object with a different address, so it represents another `specialization_id`.
Also since you are interested by the address, why do you care about "objects must be forward declarable"?
This address mechanism is different from the type approach used to name kernels and global pipes in Intel extension. Is it really simpler?

Reply:
After playing with this, it looks like the specialization_id` objects needs static storage.
But by looking at the C++20 draft, I cannot see this restriction in *13.4.2 Template non-type arguments [temp.arg.nontype]*

Some experiment in https://github.com/keryell/teaching_C-plus-plus/blob/master/NTTP_ref.cpp and https://github.com/keryell/teaching_C-plus-plus/blob/master/NTTP_ref_2.cpp